File: src/RunTests/JavaScript/runTest.js
const fs = require('fs')
const readline = require('readline')
const generateMessage = require('./messages')

/**
 * Scan the tests file and execute the tests for the given function. 
 * @param {string} testsPath - A string with the path for the test files corresponding to this test
 * @param {func} testFunction - The function which will be used to test the cases
 * @returns {void}
 */
module.exports = function readAndExecuteTests(testsPath, testFunction) {

  let input = null
  let failures = 0
  let totalNumberOfTests = undefined
  const readInterface = _generateReadInterface(testsPath)

  readInterface.on('line', function (line) {
    if (!totalNumberOfTests) {
      totalNumberOfTests = parseInt(line)
    } else if (input){
      const testPassedSuccessfully = _performTest(input, line, testFunction)
      if (!testPassedSuccessfully) failures++
      input = null
    } else {
      input = line
    }
  })

  readInterface.on('close', function () {
    generateMessage.results(totalNumberOfTests, failures)
  })

}

/**
 * 
 * @param {string} rawInput - A string with the input for the current test as received from the file reader
 * @param {string} rawOutput - A string with the output for the current test as received form the file reader
 * @param {func} testFunction - The function which will be used to test the cases
 */
function _performTest(rawInput, rawOutput, testFunction) {

  let parsedInput
  let parsedOutput

  try{
    parsedInput = JSON.parse(rawInput)
  }catch(e){
    parsedInput = rawInput
  }

  try{
    parsedOutput = JSON.parse(rawOutput)
  }catch(e){
    parsedOutput = rawOutput
  }

  const actualOutput = testFunction(parsedInput)
  const matchingOutputs = parsedOutput.toString() === actualOutput.toString()
  if (!matchingOutputs) { 
    generateMessage.failedTest(parsedInput, parsedOutput, actualOutput)
  }
  return matchingOutputs
}

/**
 * 
 * @param {string} filePath - The file to be read
 * @return { readLineInterface } a readLine interface with the file provided by the file path loaded
 */
function _generateReadInterface(filePath) {
  return readline.createInterface({
    input: fs.createReadStream(filePath),
    console: false
  })
}

File: src/RunTests/Python/run_test.py
import json
from messages import print_fail_test_message, print_results_message
from paths import get_test_file_path

"""
Entry point for the testing suite
:param function_to_test: The invoking function to test agains the test cases
:return: void
"""
def run_tests(function_to_test):
  test_file_path = get_test_file_path()
  execute_tests( test_file_path, function_to_test )

"""
Read the test file and execute tests
:param test_file_path: The path where the test file of the invoking challenge is located
:param function_to_test: The function to test agains the test file
:return: void
"""
def execute_tests(test_file_path, function_to_test):
  test_file = open(test_file_path, 'r')
  total_cases = int(test_file.readline())

  failed_tests = 0
  expected_input = None
  expected_output = None
  line = test_file.readline()
  count = 1

  while line:
    if count % 2 == 0:
      expected_output = line
      success = execute_test( expected_input, expected_output, function_to_test )
      failed_tests = failed_tests + 1 if not success else failed_tests
    else: 
      expected_input = line
    line = test_file.readline()
    count += 1

  test_file.close()
  print_results_message( total_cases, failed_tests )

"""
Execute the test read from the test file, using the given function
:param expected_input: The input for this test, as read from the test file
:param expected_output: The output for this test, as read from the test file 
:param function_to_test: The function to test with the given input
:return: bool: States if the result of the test matches the expected
"""
def execute_test( expected_input, expected_output, function_to_test ):

  parsed_input = None
  parsed_output = None

  try:
    parsed_input = json.loads( expected_input )
  except:
    parsed_input = expected_input

  try:
    parsed_output = json.loads( expected_output )
  except:
    parsed_output = expected_output

  actual_output = function_to_test( parsed_input )
  result = parsed_output == actual_output

  if not result:
    print_fail_test_message( parsed_input, parsed_output, actual_output )
  
  return result
